src/btree/bt_sync.c

```diff
@@ -143,6 +143,42 @@ __sync_ref_mark_deleted(WT_SESSION_IMPL *session, WT_REF *ref)
     return (0);
 }
 
+/*
+ * __sync_ref_int_obsolete_cleanup --
+ *     Traverse the internal page and identify the leaf pages that are obsolete and mark them as
+ *     deleted.
+ */
+static int
+__sync_ref_int_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *intref)
+{
+    WT_PAGE_INDEX *pindex;
+    WT_REF *ref;
+    uint32_t slot;
+
+    pindex = NULL;
+
+    pindex = WT_INTL_INDEX_GET_SAFE(intref->page);
+    for (slot = 0; slot < pindex->entries; slot++) {
+        /*
+         * Get a reference, setting the reference hint if it's wrong (used when we continue the
+         * walk). We don't always update the hints when splitting, it's expected for them to be
+         * incorrect in some workloads.
+         */
+        ref = pindex->index[slot];
+        if (ref->pindex_hint != slot)
+            ref->pindex_hint = slot;
+
+        /* Skip the deleted child pages. */
+        if (ref->state == WT_REF_DELETED)
+            continue;
+
+        if (__sync_ref_is_obsolete(session, ref))
+            WT_RET(__sync_ref_mark_deleted(session, ref));
```

Comment:
Doesn't this need to go recursive in order to handle internal pages that are the parents of a set of internal pages? That is, why can we assume we'll visit every internal page in the system that is a parent of a leaf page if we only visit the internal pages we encounter in a specific in-cache walk?

Reply:
The tree walk function returns all the pages that are in memory. By reading all the internal pages that are part of the history store btree even if they are not in memory will get all the internal pages one by one and in the outer function, we traverse each page for finding out the obsolete pages. This way we ensure that all the internal pages are traversed.

Reply:
I get it now, I missed what you were trying to do when I first looked at it.

I'd suggest creating a function like `__ref_is_leaf()` that returns if the page is an internal page.

Reply:
I'd also suggest `WT_READ_CACHE_LEAF` which is a subset of `WT_READ_CACHE`?

Reply:
Used `WT_READ_CACHE_LEAF`.